### DISCUSSION TITLE
Authelia OIDC Setup

### DISCUSSION THREAD

**Discussion by prologic** [authelia/authelia#5703] [General]
Hey folks,

I've been looking at integrating a few more self-hosted apps/services I have that don't support Proxy / HTTP Header Auth, but do support OIDC. Authelia is capable of being an OIDC provider, which is great.

What is not great at the moment is the amount of effort involved in setting this up. There's a lot to unpack and wrap your head around when reading the [OpenID Connect](https://www.authelia.com/configuration/identity-providers/open-id-connect/) docs provider. There's a lot to "get wrong".

Are we able to build some tools to help set this up and manage oidc clients? 🙏

**Comment by prologic** [comment]
Bump

**Reply by james-d-elliott** [reply]
Hey sorry I missed responding to this. It's been a busy few months.

I have a slight aversion to doing this as it's something that will require additional maintenance. What I do have in mind to do is to move almost all of the configuration for OpenID Connect 1.0 into the database (so we can more easily manage individual issuers rather than a flat issuer configuration, and so we can eventually handle dynamic registration). Theoretically we could add a CLI wizard which helps with the setup of a new issuer or client. It's likely we'll have a set of common default templates for applications but there are quite a few different options available.

However I am not entirely sure how much guidance we can give to users other than [specific examples](https://www.authelia.com/integration/openid-connect/) which have been tested by the community as each application is different. Do you have a more concrete idea of what you think would benefit yourself and other users?

The reason we; and almost every provider / relying party describe the options exactly as defined in the spec is to help alleviate this as an issue as it means documentation from a relying party will roughly match the client configuration for a provider and vice versa.

**Reply by prologic** [reply]
Hey @james-d-elliott 👋 Thanks for taking the time to get back to this 👌

So I think the main thing that's missing is documentation on how you generate the cryptographic material you need to construct an issuer as well as clients. I had to piece this together from several blog posts, some of which were wrong too! I completely support the direction you're taking this, centralising all the OIDC issue and client setup into the database and having a set of API(s) and even CLI(s) that use teh same is a great thing -- Really looking forward to that!

In the meanwhile, I'll try to allocate some time to see if I can help improve the docs here, maybe even a shell script _might_ be helpful too, if you're open to that?

Thanks!

**Reply by james-d-elliott** [reply]
When I look at the following sections it guides you to the relevant guides and we already have tooling to generate the keys for you. Did you miss them maybe or feel there is an inaccuracy?

- The [issuer_private_key](https://www.authelia.com/configuration/identity-providers/open-id-connect/#issuer_private_key) documentation references:
    - The key must be generated by the administrator and can be done by following the [Generating an RSA Keypair](https://www.authelia.com/reference/guides/generating-secure-values/#generating-an-rsa-keypair) guide.
- The [hmac_secret](https://www.authelia.com/configuration/identity-providers/open-id-connect/#hmac_secret) documentation references:
    - It’s strongly recommended this is a [Random Alphanumeric String](https://www.authelia.com/reference/guides/generating-secure-values/#generating-a-random-alphanumeric-string) with 64 or more characters.

**Reply by prologic** [reply]
Oh gawd 😱 I guess I did?! For some reason I never noticed, I got overwhelmed by the complexity of what was in front of me! That's my fault.

**Reply by james-d-elliott** [reply]
It happens I am curious if you know why you missed it if you know and how we're reduce the chance of that though?

**Reply by prologic** [reply]
By the way @james-d-elliott -- Just curious what the timeline/roadmap around having a Web administration interface for Authelia and a database backing things like rules, users, groups and oidc clients? 🤔
